### PR TITLE
[Wasm64] Run core tests under wasm64 as well as wasm64l

### PR DESCRIPTION
**Pull request by sbc100**

### CODE REVIEW

**Comment by sbc100** [comment]
Because wasm64 tests only run under d8 this requires some fixes of things that were just related to being able to run stuff under d8.

File: .circleci/config.yml [hunk 1]
```diff
@@ -449,7 +449,10 @@ jobs:
     steps:
       - run-tests-linux:
           frozen_cache: false
-          test_targets: "wasm64.test_hello_world wasm64.test_ccall wasm64l"
+          test_targets: "wasm64"
+      - run-tests-linux:
+          frozen_cache: false
+          test_targets: "wasm64l"
```
Comment (kripken) [inline .circleci/config.yml:455]: Do we need wasm64l now that we have full wasm64 up and running?

File: test/test_core.py [hunk 2]
```diff
@@ -5873,8 +5877,12 @@ def test_fs_mmap(self):
     self.uses_es6 = True
     for fs in ['MEMFS', 'NODEFS', 'NODERAWFS']:
       if fs == 'NODEFS':
+        if not self.is_node():
```
Comment (kripken) [inline test/test_core.py:5880]: This checks if there is 1 JS engine and it is node, but what if the list of engines is `[node, other]`?

Comment (sbc100) [reply test/test_core.py:5880]: That would fail since the non-node engine could not run the code.   This part of the test requires that node (and only node) is used to run the test.

Running with more than one engine configured is very rare.. we have no testing for that configuration that I know of.  Do you often do it?  What would you expect to happen in that case?   Perhaps these tests should narrow the list, like `require_node` does?

File: .circleci/config.yml (hunk 1 above)
Comment (sbc100) [reply .circleci/config.yml:455]: Yes, I think wasm64l will always be useful won't it?  IIRC wouter designed to mode mostly so that 64-bit software could be run on wasm without requiring it to be ported to IPL32... In that use case such software still wants to target wasm32.. assuming it doesn't need more than 4Gb of memory.

Comment (kripken) [reply .circleci/config.yml:455]: Hmm, true... but I thought the main use case was to bootstrap testing on wasm64, which is now complete basically. For the use case of helping people port 64-bit code to wasm32, I'm not sure it's worth us running all of `wasm64l` constantly here. We do test the lowering pass in Binaryen, and we could test just a select few tests here perhaps. That seems more efficient to me. But we don't need to decide on that urgently I guess.

File: test/test_core.py (hunk 2 above)
Comment (kripken) [reply test/test_core.py:5880]: I think we have `--all-engines` for multi-engine testing? I don't do it often since engine bugs are so rare, though.

Narrowing the list seems better to me, good idea.

Another option might be to add code in a single place that does that narrowing automatically if `NODEFS` is in the commandline arguments?

File: .circleci/config.yml (hunk 1 above)
Comment (sbc100) [reply .circleci/config.yml:455]: The other issue is that wasm64 is only currently runnable under v8..  we need to keep wasm64l around at least until we can run wam64 under node and in the browser.   Also, these tests really don't take that long to run since they are still a subset of `core`.   Each of these runs takes around 2 minutes in CI

File: test/test_core.py (hunk 2 above)
Comment (sbc100) [reply test/test_core.py:5880]: > I think we have `--all-engines` for multi-engine testing? I don't do it often since engine bugs are so rare, though.
> 

The problem with `--all-engines` is that we don't test it, so at any given time is likely broken.

I wonder if its worth removing it for this reason?  What are the chances that the entire `other` or `core` test suite will run will `--all-engines` set?

Having said that i think can do the narrowing thing here..

**Comment by dschuff** [comment]
Does wasm64 not run under Chrome even with a flag? I thought it was possible to pass arbitrary JS flags through to V8?

File: test/test_core.py (hunk 2 above)
Comment (kripken) [reply test/test_core.py:5880]: > What are the chances that the entire other or core test suite will run will --all-engines set?

Maybe the chances aren't high, since we don't test it. I guess it's probably fine at this point to remove all-engines - we're committed to running most of the tests in node, and only a select few in v8.

Comment (sbc100) [reply test/test_core.py]: I remove this part of the change.

**Comment by sbc100** [comment]
> Does wasm64 not run under Chrome even with a flag? I thought it was possible to pass arbitrary JS flags through to V8?

It most likely does but we want but it would be a shame to have to more all the wasm64 testing into browser tests.. and we don't currently run the core tests in the browser (nor do we want to start doing that).

**Comment by sbc100** [comment]
I think we should land this since it fixes a bunch of issues, especially in the test code.  We can debate removing wasm64l seperately/later.  This only adds 2 minutes to an already fast builder as it stands.

File: test/test_core.py [hunk 3]
```diff
@@ -6542,6 +6559,8 @@ def test_neon_wasm_simd(self):
   @no_safe_heap('has unaligned 64-bit operations in wasm')
   @no_ubsan('test contains UB')
   def test_sse1(self):
+    if self.is_wasm64():
+      self.skipTest('https://github.com/llvm/llvm-project/issues/57577')
```
Comment (dschuff) [inline test/test_core.py:6563]: why the explicit self.skipTest in some places and @no_wasm64 in others?

File: test/test_core.py [hunk 4]
```diff
@@ -2101,6 +2103,8 @@ def test_emscripten_get_compiler_setting(self):
     self.do_runf(src, read_file(output).replace('waka', shared.EMSCRIPTEN_VERSION))
 
   def test_emscripten_has_asyncify(self):
+    if self.get_setting('MEMORY64') == 1:
```
Comment (dschuff) [inline test/test_core.py:2106]: is this different from is_wasm64?

**Review by dschuff — COMMENTED** [review]
agreed that we can go ahead and get this tested ASAP, independently of wasm64l.

**Comment by dschuff** [comment]
BTW the "Current" version of node (18.9.0) says it has support for the --experimental-wasm-memory64 flag, so that might be an option too.

File: test/test_core.py [hunk 5]
```diff
@@ -6095,6 +6105,10 @@ def test_unistd_io(self):
       self.emcc_args = orig_compiler_opts + ['-D' + fs]
       if fs == 'NODEFS':
         self.emcc_args += ['-lnodefs.js']
+        if config.NODE_JS not in config.JS_ENGINES:
+          # NODEFS requires node
+          continue
```
Comment (kripken) [inline test/test_core.py:6110]: Why is this needed here but not in the diff above it that just uses `require_node`?

File: test/test_core.py (hunk 3 above)
Comment (sbc100) [reply test/test_core.py:6563]: Hmm.. I think that must have been the result of some intermediate step.  Fixed.

File: test/test_core.py (hunk 5 above)
Comment (sbc100) [reply test/test_core.py:6110]: `require_node` will call `skipTest` it can't be used to skip over just one loop iteration like this;  It would skip the entire rest of the test.

Ideally we would avoid loops in tests for this reason, but I wasn't able convert this test to parameterized form.

**Review by kripken — APPROVED** [review]

**Comment by tiran** [comment]
> BTW the "Current" version of node (18.9.0) says it has support for the --experimental-wasm-memory64 flag, so that might be an option too.

I have been using Node 16.14.0 for wasm64 with great success.
